include/boost/core/lightweight_test.hpp

```diff
@@ -118,106 +120,76 @@ template<class T> inline const void* test_output_impl(T volatile* v) { return co
+    static const char* op() { return "!="; }
+    static const char* anti_op() { return "=="; }
+};
 
-template<class T, class U> inline void test_le_impl( char const * expr1, char const * expr2,
-  char const * file, int line, char const * function, T const & t, U const & u )
-{
-    if( t <= u )
-    {
-        report_errors_remind();
-    }
-    else
-    {
-        BOOST_LIGHTWEIGHT_TEST_OSTREAM
-            << file << "(" << line << "): test '" << expr1 << " <= " << expr2
-            << "' failed in function '" << function << "': "
-            << "'" << test_output_impl(t) << "' > '" << test_output_impl(u) << "'" << std::endl;
-        ++test_errors();
-    }
-}
+template <typename T>
+struct test_output_traits<std::greater<T> > {
+    static const char* op() { return ">"; }
+    static const char* anti_op() { return "<="; }
+};
+
+template <typename T>
+struct test_output_traits<std::less<T> > {
+    static const char* op() { return "<"; }
+    static const char* anti_op() { return ">="; }
+};
+
+template <typename T>
+struct test_output_traits<std::greater_equal<T> > {
+    static const char* op() { return ">="; }
+    static const char* anti_op() { return "<"; }
+};
 
-template<class T, class U> inline void test_gt_impl( char const * expr1, char const * expr2,
-  char const * file, int line, char const * function, T const & t, U const & u )
+template <typename T>
+struct test_output_traits<std::less_equal<T> > {
+    static const char* op() { return "<="; }
+    static const char* anti_op() { return ">"; }
+};
+
+template <typename T>
+struct test_output_traits<boost::core::close_to<T> > {
+    static const char* op() { return "~="; }
+    static const char* anti_op() { return "!="; }
+};
+
+template<class T, class U, class BinaryPredicate>
+inline void test_with_impl(char const * expr1, char const * expr2,
+                           char const * file, int line, char const * function,
+                           T const & t, U const & u, BinaryPredicate pred)
 {
-    if( t > u )
+    if( pred(t, u) )
     {
         report_errors_remind();
     }
     else
     {
         BOOST_LIGHTWEIGHT_TEST_OSTREAM
-            << file << "(" << line << "): test '" << expr1 << " > " << expr2
+            << file << "(" << line << "): test '" << expr1 << " " << test_output_traits<BinaryPredicate>::op() << " " << expr2
             << "' failed in function '" << function << "': "
-            << "'" << test_output_impl(t) << "' <= '" << test_output_impl(u) << "'" << std::endl;
+            << "'" << test_output_impl(t) << "' " << test_output_traits<BinaryPredicate>::anti_op() << " '" << test_output_impl(u) << "'" << std::endl;
         ++test_errors();
     }
 }
 
-template<class T, class U> inline void test_ge_impl( char const * expr1, char const * expr2,
-  char const * file, int line, char const * function, T const & t, U const & u )
+template<template <class> class BinaryPredicate, class T, class U>
```

Comment:
There's also no need to confine the predicates to only templates of the form `template<class>`